extensions/amp-gwd-animation/0.1/amp-gwd-animation.js

```diff
@@ -0,0 +1,228 @@
+ * GWD runtime service, which is invoked with the evaluated arguments
+ * (@see getActionImplArgs and createAction_).
+ * @const {!Object<string, !Array<string>>}
+ */
+const ACTION_IMPL_ARGS = {
+  'play': ['args.id'],
+  'pause': ['args.id'],
+  'togglePlay': ['args.id'],
+  'gotoAndPlay': ['args.id', 'args.label'],
+  'gotoAndPause': ['args.id', 'args.label'],
+  'gotoAndPlayNTimes': ['args.id', 'args.label', 'args.N', 'event.eventName'],
+  'setCurrentPage': ['args.index'],
+};
+
+/**
+ * Given an action name, extracts its required arguments from an invocation
+ * payload, which may contain the data needed by this action in its `args` or
+ * `events` child objects. The arguments required by each action are specified
+ * in ACTION_IMPL_ARGS as path strings.
+ * @param {string} actionName
+ * @param {!../../../src/service/action-impl.ActionInvocation} invocation
+ * @return {!Array<string>}
+ */
+const getActionImplArgs = function(actionName, invocation) {
+  const argDefs = user().assert(
+      ACTION_IMPL_ARGS[actionName],
+      `The action ${actionName} is not a supported action.`);
+
+  return argDefs.map(argDef => {
+    // Walk the invocation object to get the requested property by
+    // its path, e.g., 'args.id' == invocation.args.id
+    let obj = invocation;
+    const props = argDef.split('.');
+    for (let i = 0; i < props.length; i++) {
+      if (!obj) {
+        return undefined;
+      }
+      obj = obj[props[i]];
+    }
+
+    return obj;
+  });
+};
+
+export class GwdAnimation extends AMP.BaseElement {
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /**
+     * The prefix to use for triggered timeline events if it is necessary to
+     * avoid event name conflicts, supplied via the `timeline-event-prefix`
+     * attribute (@see buildCallback).
+     * @private {string}
+     */
+    this.timelineEventPrefix_ = '';
+
+    /** @private {!Function} */
+    this.boundOnGwdTimelineEvent_ = this.onGwdTimelineEvent_.bind(this);
+  }
+
+  /** @override */
+  buildCallback() {
+    // TODO(sklobovskaya): Remove experiment guard.
+    user().assert(
+        isExperimentOn(this.getAmpDoc().win, EXPERIMENT),
+        `Experiment ${EXPERIMENT} is disabled.`);
+
+    // Ensure GWD animation runtime service factory is registered for this
+    // ampdoc. In the event multiple copies of this extension exist in the doc,
+    // the runtime service will only be started once.
+    // TODO(sklobovskaya): This can be moved out elsewhere (e.g., to the
+    // extension registration callback) when the required registration APIs are
+    // ready.
+    installGwdRuntimeServiceForDoc(this.getAmpDoc());
+
+    this.timelineEventPrefix_ =
+        this.element.getAttribute('timeline-event-prefix') || '';
+
+    // Listen for GWD timeline events to re-broadcast them via the doc action
+    // service.
+    this.getAmpDoc().win.addEventListener(
```

Comment:
Same question here. Is it possible to catch events on the root node instead?

Reply:
Done.